### PR TITLE
refactor!: move useBeamSpotConstraint to vertexingOptions

### PR DESCRIPTION
**Pull request by felix-russo**
Previously, the boolean member variable `useBeamSpotConstraint` of the AMVF and IVF was set independently of the actual beam spot constraint which was part of the `vertexingOptions`. This PR moves `useBeamSpotConstraint` to `vertexingOptions` as well.

### CODE REVIEW

**Comment by paulgessinger** [comment]
I guess this is breaking?

**Comment by paulgessinger** [comment]
No not at all, we just have to handle it correctly. I might have to ask you to do some checks on the Athena side to determine the impact (cc @CarloVarni @noemina)

File: Core/include/Acts/Vertexing/AdaptiveMultiVertexFinder.hpp [hunk 1]
```diff
@@ -318,12 +316,13 @@ class AdaptiveMultiVertexFinder {
...
   /// @param seedTracks The seed tracks
   /// @param fitterState The vertex fitter state
+  /// @param useBeamSpotConstraint Indicates whether beam spot constrain is used
   ///
   /// @return pair(nCompatibleTracks, isGoodVertex)
   std::pair<int, bool> checkVertexAndCompatibleTracks(
       Vertex<InputTrack_t>& vtx,
       const std::vector<const InputTrack_t*>& seedTracks,
-      FitterState_t& fitterState) const;
+      FitterState_t& fitterState, const bool& useBeamSpotConstraint) const;
```
Comment (paulgessinger) [inline Core/include/Acts/Vertexing/AdaptiveMultiVertexFinder.hpp:325]: The reference is larger than the `bool`, so I think a by-value `bool` argument is fine.

```suggestion
      FitterState_t& fitterState, bool useBeamSpotConstraint) const;
```

File: Core/include/Acts/Vertexing/VertexingOptions.hpp [hunk 2]
```diff
@@ -24,22 +24,44 @@ struct VertexingOptions {
...
-  VertexingOptions(
-      const GeometryContext& gctx, const MagneticFieldContext& mctx,
-      const Vertex<input_track_t>& vconstr = Vertex<input_track_t>())
-      : geoContext(gctx), magFieldContext(mctx), vertexConstraint(vconstr) {}
+  /// @param bsConstr The beamspot constraint
+  /// @param useConstr Boolean indicating whether vertex constraint should be used
+  VertexingOptions(const GeometryContext& gctx,
+                   const MagneticFieldContext& mctx,
+                   const Vertex<input_track_t>& bsConstr,
+                   const bool useConstr = true)
```
Comment (paulgessinger) [inline Core/include/Acts/Vertexing/VertexingOptions.hpp:36]: Could this be something like

```suggestion
                   std::optional<Vertex<input_track_t>> bsConstr)
```

Maybe we could get rid of the other constructors in that case.

File: Core/include/Acts/Vertexing/VertexingOptions.hpp [hunk 3]
```diff
@@ -24,22 +24,44 @@ struct VertexingOptions {
...
+      : geoContext(gctx),
+        magFieldContext(mctx),
+        beamSpotConstraint(bsConstr),
+        useBeamSpotConstraint(useConstr) {
+    assert((!useBeamSpotConstraint ||
```
Comment (paulgessinger) [inline Core/include/Acts/Vertexing/VertexingOptions.hpp:41]: Should this be checked even when not in `DEBUG`?

**Review by paulgessinger — COMMENTED** [review]
Here are a few comments.

Can we get this in before the end of the week in time for `v28`?

File: Core/include/Acts/Vertexing/VertexingOptions.hpp (hunk 3 above)
Comment (felix-russo) [reply Core/include/Acts/Vertexing/VertexingOptions.hpp]: Like this?

File: Core/include/Acts/Vertexing/VertexingOptions.hpp (hunk 2 above)
Comment (felix-russo) [reply Core/include/Acts/Vertexing/VertexingOptions.hpp:36]: Unfortunately, the beam spot constraint is always used as the initial guess of the vertex position. Therefore, it needs to be set!
I guess the name of the boolean is kind of misleading, should I change it? If so, do you have an idea?

File: Core/include/Acts/Vertexing/VertexingOptions.hpp (hunk 3 above)
Comment (paulgessinger) [reply Core/include/Acts/Vertexing/VertexingOptions.hpp]: I was just trying to say that `assert` only executes in `DEBUG` builds. Should we check this consistency in all case to be sure?

**Review by paulgessinger — DISMISSED** [review]

**Comment by paulgessinger** [comment]
The bridged physmon job fails due (at least) 3 FPEs that are encountered. From the `ckf_tracking` log:

```
17:41:39    Sequencer      INFO      - FLTUND: (3 times) 
 0# Acts::AdaptiveMultiVertexFinder<Acts::AdaptiveMultiVertexFitter<Acts::GenericBoundTrackParameters<Acts::SinglyCharged>, Acts::HelicalTrackLinearizer<Acts::Propagator<Acts::EigenStepper<Acts::StepperExtensionList<Acts::detail::GenericDefaultExtension<double> >, Acts::detail::VoidAuctioneer>, Acts::detail::VoidNavigator>, Acts::PropagatorOptions<Acts::ActionList<>, Acts::AbortList<> > > >, Acts::TrackDensityVertexFinder<Acts::AdaptiveMultiVertexFitter<Acts::GenericBoundTrackParameters<Acts::SinglyCharged>, Acts::HelicalTrackLinearizer<Acts::Propagator<Acts::EigenStepper<Acts::StepperExtensionList<Acts::detail::GenericDefaultExtension<double> >, Acts::detail::VoidAuctioneer>, Acts::detail::VoidNavigator>, Acts::PropagatorOptions<Acts::ActionList<>, Acts::AbortList<> > > >, Acts::GaussianTrackDensity<Acts::GenericBoundTrackParameters<Acts::SinglyCharged> > > >::find(std::vector<Acts::GenericBoundTrackParameters<Acts::SinglyCharged> const*, std::allocator<Acts::GenericBoundTrackParameters<Acts::SinglyCharged> const*> > const&, Acts::VertexingOptions<Acts::GenericBoundTrackParameters<Acts::SinglyCharged> > const&, Acts::AdaptiveMultiVertexFinder<Acts::AdaptiveMultiVertexFitter<Acts::GenericBoundTrackParameters<Acts::SinglyCharged>, Acts::HelicalTrackLinearizer<Acts::Propagator<Acts::EigenStepper<Acts::StepperExtensionList<Acts::detail::GenericDefaultExtension<double> >, Acts::detail::VoidAuctioneer>, Acts::detail::VoidNavigator>, Acts::PropagatorOptions<Acts::ActionList<>, Acts::AbortList<> > > >, Acts::TrackDensityVertexFinder<Acts::AdaptiveMultiVertexFitter<Acts::GenericBoundTrackParameters<Acts::SinglyCharged>, Acts::HelicalTrackLinearizer<Acts::Propagator<Acts::EigenStepper<Acts::StepperExtensionList<Acts::detail::GenericDefaultExtension<double> >, Acts::detail::VoidAuctioneer>, Acts::detail::VoidNavigator>, Acts::PropagatorOptions<Acts::ActionList<>, Acts::AbortList<> > > >, Acts::GaussianTrackDensity<Acts::GenericBoundTrackParameters<Acts::SinglyCharged> > > >::State&) const [clone .isra.0] at /builds/acts/ci-bridge/src/Core/include/Acts/Vertexing/AdaptiveMultiVertexFinder.ipp:120
 1# ActsExamples::ProcessCode ActsExamples::AdaptiveMultiVertexFinderAlgorithm::executeAfterSeederChoice<Acts::TrackDensityVertexFinder<Acts::AdaptiveMultiVertexFitter<Acts::GenericBoundTrackParameters<Acts::SinglyCharged>, Acts::HelicalTrackLinearizer<Acts::Propagator<Acts::EigenStepper<Acts::StepperExtensionList<Acts::detail::GenericDefaultExtension<double> >, Acts::detail::VoidAuctioneer>, Acts::detail::VoidNavigator>, Acts::PropagatorOptions<Acts::ActionList<>, Acts::AbortList<> > > >, Acts::GaussianTrackDensity<Acts::GenericBoundTrackParameters<Acts::SinglyCharged> > >, Acts::AdaptiveMultiVertexFinder<Acts::AdaptiveMultiVertexFitter<Acts::GenericBoundTrackParameters<Acts::SinglyCharged>, Acts::HelicalTrackLinearizer<Acts::Propagator<Acts::EigenStepper<Acts::StepperExtensionList<Acts::detail::GenericDefaultExtension<double> >, Acts::detail::VoidAuctioneer>, Acts::detail::VoidNavigator>, Acts::PropagatorOptions<Acts::ActionList<>, Acts::AbortList<> > > >, Acts::TrackDensityVertexFinder<Acts::AdaptiveMultiVertexFitter<Acts::GenericBoundTrackParameters<Acts::SinglyCharged>, Acts::HelicalTrackLinearizer<Acts::Propagator<Acts::EigenStepper<Acts::StepperExtensionList<Acts::detail::GenericDefaultExtension<double> >, Acts::detail::VoidAuctioneer>, Acts::detail::VoidNavigator>, Acts::PropagatorOptions<Acts::ActionList<>, Acts::AbortList<> > > >, Acts::GaussianTrackDensity<Acts::GenericBoundTrackParameters<Acts::SinglyCharged> > > > >(ActsExamples::AlgorithmContext const&, Acts::TrackDensityVertexFinder<Acts::AdaptiveMultiVertexFitter<Acts::GenericBoundTrackParameters<Acts::SinglyCharged>, Acts::HelicalTrackLinearizer<Acts::Propagator<Acts::EigenStepper<Acts::StepperExtensionList<Acts::detail::GenericDefaultExtension<double> >, Acts::detail::VoidAuctioneer>, Acts::detail::VoidNavigator>, Acts::PropagatorOptions<Acts::ActionList<>, Acts::AbortList<> > > >, Acts::GaussianTrackDensity<Acts::GenericBoundTrackParameters<Acts::SinglyCharged> > > const&) const at /builds/acts/ci-bridge/src/Examples/Algorithms/Vertexing/src/AdaptiveMultiVertexFinderAlgorithm.cpp:161
 2# ActsExamples::AdaptiveMultiVertexFinderAlgorithm::execute(ActsExamples::AlgorithmContext const&) const at /builds/acts/ci-bridge/src/Examples/Algorithms/Vertexing/src/AdaptiveMultiVertexFinderAlgorithm.cpp:65
 3# ActsExamples::Sequencer::run()::{lambda()#1}::operator()() const::{lambda(tbb::detail::d1::blocked_range<unsigned long> const&)#1}::operator()(tbb::detail::d1::blocked_range<unsigned long> const&) const at /builds/acts/ci-bridge/src/Examples/Framework/src/Framework/Sequencer.cpp:505
 4# tbb::detail::d1::start_for<tbb::detail::d1::blocked_range<unsigned long>, ActsExamples::Sequencer::run()::{lambda()#1}::operator()() const::{lambda(tbb::detail::d1::blocked_range<unsigned long> const&)#1}, tbb::detail::d1::auto_partitioner const>::execute(tbb::detail::d1::execution_data&) at /usr/include/oneapi/tbb/parallel_for.h:172
 5# 0x00007FA411E61B3C in /usr/lib/x86_64-linux-gnu/libtbb.so.12
 6# 0x00007FA411E63DD8 in /usr/lib/x86_64-linux-gnu/libtbb.so.12
 7# 0x00007FA413C07B43 in /usr/lib/x86_64-linux-gnu/libc.so.6
```

Any ideas @felix-russo?

**Review by CarloVarni — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results
### :red_circle: Some tests have failed!
Please investigate the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/6119076)!

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058466) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058465) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058464) |
|:red_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058463) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058462) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058461) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058460) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058459) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058458) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058457) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058456) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058455) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058454) |
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058453) |
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058452) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058451) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/32058450) |

**Comment by paulgessinger** [comment]
Transient error, breaks (broke) the build though.

**Pull request merged**